tuist/tuist
# Fix warm caching failing due to a wrong assumption around Mac Catalyst

### Short description 📝
While testing the cache, @Alex-Ozun noted that Tuist yielded "X not found" with a Swift Package integrated through `Dependencies.swift`. After some debugging, it turns out that our warming logic assumed that the only condition to build for  Mac Catalyst is **having some** frameworks that support Catalyst. This is assumption has proven invalid. In his case, which can easily be reproduced [here](https://github.com/Alex-Ozun/TuistCacheRepro), the package dependency supports Mac Catalyst but the dependent target not, and that caused the issue.

To fix it, I extended the logic to traverse the graph when determining if a target can be built for Mac Catalyst. In the future, we might want to iterate on the logic to make it more granular, but that'll come at the cost of some complexity in the logic that generates the schemes because we no longer can have a large scheme that aggregates all the binary targets together.

**Note** that the logic to determine if a dependency supports Mac Catalyst only works with source targets. In another PR I'll extend the logic to obtain that information from pre-compiled binaries such as libraries, frameworks, and xcframeworks.

### How to test the changes locally 🧐
You should be able to successfully run `swift run tuist cache warm -x` against [this project](https://github.com/Alex-Ozun/TuistCacheRepro) and then generate a project passing the `-x` flag and get the dependency as a binary and a project that compiles.

### Contributor checklist ✅


### Reviewer checklist ✅
